File: myFinance_improved_V2/lib/features/inventory_analysis/discrepancy/domain/entities/discrepancy_overview.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'discrepancy_overview.freezed.dart';

/// Store inventory discrepancy data
@freezed
class StoreDiscrepancy with _$StoreDiscrepancy {
  const StoreDiscrepancy._();

  const factory StoreDiscrepancy({
    required String storeId,
    required String storeName,
    required int totalEvents,
    required int increaseCount,
    required int decreaseCount,
    required num increaseValue,
    required num decreaseValue,
    required num netValue,
    String? status, // 'abnormal', 'warning', 'normal' (statistical test result)
  }) = _StoreDiscrepancy;

  /// Increase/decrease ratio (increase count / total count)
  num get increaseRatio =>
      totalEvents > 0 ? increaseCount / totalEvents : 0;

  /// Net ratio (net value / decrease value)
  num? get netRatio =>
      decreaseValue != 0 ? netValue / decreaseValue * 100 : null;

  /// Status label
  String get statusLabel {
    return switch (status) {
      'abnormal' => 'Statistical anomaly',
      'warning' => 'Warning',
      _ => 'Normal range',
    };
  }
}

/// Inventory discrepancy aggregate data
/// RPC: get_discrepancy_overview response
@freezed
class DiscrepancyOverview with _$DiscrepancyOverview {
  const DiscrepancyOverview._();

  const factory DiscrepancyOverview({
    required String status, // 'ok', 'insufficient_data'
    String? message,
    String? minRequired,
    required num totalIncreaseValue,
    required num totalDecreaseValue,
    required num netValue,
    required int totalStores,
    required int totalEvents,
    required List<StoreDiscrepancy> stores,
  }) = _DiscrepancyOverview;

  /// Whether data is insufficient
  bool get isInsufficientData => status == 'insufficient_data';

  /// Whether analysis is possible
  bool get canAnalyze => status == 'ok';

  /// Net ratio
  num? get netRatio {
    if (totalDecreaseValue == 0) return null;
    return netValue / totalDecreaseValue * 100;
  }

  /// Overall status (when analysis is possible)
  String get analysisStatus {
    if (isInsufficientData) return 'insufficient';

    // Net loss >= 10% is critical
    if (netRatio != null && netRatio! < -10) return 'critical';
    // Net loss >= 5% is warning
    if (netRatio != null && netRatio! < -5) return 'warning';
    return 'good';
  }

  /// Abnormal store count
  int get abnormalStoreCount =>
      stores.where((s) => s.status == 'abnormal').length;
}

File: website/src/features/inventory_analysis/domain/entities/discrepancyOverview.ts
/**
 * Discrepancy Overview Entity
 * Domain entity for RPC response mapping
 */

// Cumulative totals
export interface DiscrepancyCumulative {
  increaseValue: number;
  decreaseValue: number;
  netValue: number;
  netPct: number;
}

// Store discrepancy data (success case)
export interface DiscrepancyStore {
  storeName: string;
  netValue: number;
  increaseValue: number;
  decreaseValue: number;
  increaseCount: number;
  decreaseCount: number;
}

// Store data for insufficient case
export interface InsufficientStore {
  storeName: string;
  totalEvents: number;
  netValue: number;
}

// Success response
export interface DiscrepancyOverviewSuccess {
  status: 'success';
  period: string;
  cumulative: DiscrepancyCumulative;
  stores: DiscrepancyStore[];
}

// Insufficient data response
export interface DiscrepancyOverviewInsufficient {
  status: 'insufficient_data';
  message: string;
  minRequired: string;
  stores: InsufficientStore[] | null;
}

// Union type for all responses
export type DiscrepancyOverview = DiscrepancyOverviewSuccess | DiscrepancyOverviewInsufficient;

/**
 * Maps RPC response to domain entity
 */
export function mapDiscrepancyOverviewFromRpc(data: Record<string, unknown>): DiscrepancyOverview {
  const status = data.status as string;

  if (status === 'insufficient_data') {
    const storesData = data.stores as Array<Record<string, unknown>> | null;

    return {
      status: 'insufficient_data',
      message: String(data.message ?? ''),
      minRequired: String(data.min_required ?? ''),
      stores: storesData
        ? storesData.map((store) => ({
            storeName: String(store.store_name ?? ''),
            totalEvents: Number(store.total_events ?? 0),
            netValue: Number(store.net_value ?? 0),
          }))
        : null,
    };
  }

  // Success case
  const cumulativeData = data.cumulative as Record<string, unknown> | undefined;
  const storesData = data.stores as Array<Record<string, unknown>> | undefined;

  return {
    status: 'success',
    period: String(data.period ?? ''),
    cumulative: {
      increaseValue: Number(cumulativeData?.increase_value ?? 0),
      decreaseValue: Number(cumulativeData?.decrease_value ?? 0),
      netValue: Number(cumulativeData?.net_value ?? 0),
      netPct: Number(cumulativeData?.net_pct ?? 0),
    },
    stores: (storesData ?? []).map((store) => ({
      storeName: String(store.store_name ?? ''),
      netValue: Number(store.net_value ?? 0),
      increaseValue: Number(store.increase_value ?? 0),
      decreaseValue: Number(store.decrease_value ?? 0),
      increaseCount: Number(store.increase_count ?? 0),
      decreaseCount: Number(store.decrease_count ?? 0),
    })),
  };
}
